### ISSUE TITLE
Switch .env file based on NODE_ENV

### ISSUE THREAD

**Issue by matthewmueller** [prisma/prisma#3865]
**This issue depends on**: https://github.com/prisma/prisma/issues/3720.

## Problem

We don't offer a native solution for switching environments when running the CLI. A good illustration of this problem is if you want to migrate your test database, you need to run prisma migrate with your test environment.

The current workaround is to either provide the environment variables explicitly:

```
DATABASE_URL="db_test_url" prisma migrate
```

Or call your environment loader from the outside:

```
dotenv -e env.test -- prisma migrate
```

## Proposal

Blitz, Next.js, Create React App support `.env.*`. This approach is nicely summed up on Next.js's documentation

> In general only one .env.local file is needed. However, sometimes you might want to add some defaults for the development (next dev) or production (next start) environment.
>
> Next.js allows you to set defaults in .env (all environments), .env.development (development environment), and .env.production (production environment).
>
> .env.local always overrides the defaults set.
>
> Note: .env, .env.development, and .env.production files should be included in your repository as they define defaults. .env*.local should be added to .gitignore, as those files are intended to be ignored. .env.local is where secrets can be stored.

Source: https://nextjs.org/docs/basic-features/environment-variables#default-environment-variables

### Tradeoffs

We're hesitant to implement this because:

- It's a very specific solution to environment management. There's a risk that this could change over time.
- Adds some runtime complexity to the client
- We want to make sure we support other environment loaders. For example, https://github.com/99designs/aws-vault.

## Related Issues and Discussion

- https://github.com/prisma/prisma/issues/3720#issuecomment-704301413
- https://github.com/prisma/prisma/issues/1717

**Comment by flybayer** [comment]
Perhaps this could be an opt-in feature with some type of config flag. I would be totally fine with that.

**Comment by NMinhNguyen** [comment]
Would be great if `.env.local` was supported out of the box.

**Comment by PhilippMolitor** [comment]
Are there any plans on supporting use of existing configuration from code?
I bundle all my config in a `config.ts` module which loads values from various environment variables, sets defaults and also handles environments. I would really like to use that with Prisma.

**Comment by flybayer** [comment]
Duplicate issue: https://github.com/prisma/prisma/issues/3226

**Comment by flybayer** [comment]
I've come to the conclusion the core issue here is that Prisma doesn't have a way to let us load our own env variables.

Currently we only have two options:
- load env variables before invoking prisma
- use `.env` or `.env.local`

But we are missing a third option which is a hook from prisma that lets us load our own environment variables.

So instead of 
```bash
dotenv-flow prisma migrate dev
```
we can just do
```bash
prisma migrate dev
```

**At the most basic level, prisma can accept a path to a shell script that we can then use to load env variables however we want, including dotenv-flow.**

That said, dotenv-flow is so common in the JS ecosystem with almost every framework supporting it that I still want an easy way to enable dotenv-flow. Perhaps via `"dotenv-flow":true` in the prisma package.json config and installing dotenv-flow as a peer dependency to prisma.

**Comment by PhilippMolitor** [comment]
maybe a duplicate, but with a more flexible approach: #4798

**Comment by matthewmueller** [comment]
Let me try to narrow the problem scope a bit and see if there are still issues:

**1. `dotenv` support is not really needed for running your application (with the Prisma Client) in production because the production environment is provided by the host (e.g. Heroku, Lambda, Vercel)** 

Are there any cases where this isn't true and you still need a `.env.production`? Maybe VMs?

**2. You need to be able to switch the environment when running CLI commands like `prisma migrate`** 

Sometimes you want to migrate your test or production database from your development machine. Are there other cases?

---

If this makes sense so far, the part I don't understand is that `prisma migrate dev` by itself cannot know to switch environments for migrating the test or production database. You need to indicate that somehow, maybe:

* `NODE_ENV=production prisma migrate dev` **or** 
* `prisma migrate dev --env=production`

So if you need to switch environments somehow, how important is it that the command is:

* `NODE_ENV=production prisma migrate dev` **instead of**
* `dotenv -e .env.production prisma migrate dev`

Given that most of the time you stick these commands in an NPM script or Makefile:

```json
{
  "pro.migrate": "NODE_ENV=production prisma migrate dev"
}
```

Or if you're lucky, the framework provides it `rails db:migrate RAILS_ENV=production`.

**Comment by flybayer** [comment]
> **1. `dotenv` support is not really needed for running your application (with the Prisma Client) in production because the production environment is provided by the host (e.g. Heroku, Lambda, Vercel)**
> 
> Are there any cases you're facing where this isn't true and you still need a `.env.production`? Maybe VMs?

As far as prisma is concerned, you are correct. Because the production DB is set via host env settings. However we do use `.env` files in production for several uses:

Next.js only allows `NODE_ENV` to be `development` or `production`, so there has to be another way to manage "app environment". The standard way to do this with Next.js/Blitz apps is via `APP_ENV`

- `.env` has `APP_ENV=development`
- `.env.production` has `APP_ENV=production`
- `.env.test` has `APP_ENV=test`

Those are all checked into git. Also, we store various other environment variables in `.env`, `.env.development`, `.env.test`, and `.env.production` that are not sensitive but needed by developers and production.

> **2. You need to be able to switch the environment when running CLI commands like `prisma migrate`**
> 
> Sometimes you want to migrate your test or production database from your development machine. Are there other cases you're running into?

Yes
 
> If this makes sense so far, the part I don't understand is that `prisma migrate dev` by itself cannot know to switch environments for migrating the test or production database. You need to indicate that somehow, maybe:
> 
> * `prisma migrate dev --env .env.production`

First problem is that env vars need loaded from both `.env` and `.env.production` in this case and merged together.

> So if you need to switch environments somehow, how important is it that the command is:
> 
> * `prisma migrate dev --env .env.production` **instead of**
> * `dotenv -e .env.production prisma migrate dev`
> 
> Given that most of the time you stick these commands in an NPM script or Makefile:
> 
> ```json
> {
>   "scripts": {
>     "pro.migrate": "dotenv -e .env.production prisma migrate dev"
>   }
> }
> ```

I don't want either one of these because usually I _don't_ put them in a package json script :) I only put them there if they are complex. The goal is that they are very simple, like `prisma migrate dev`. 
Also you need to use the other commands too and so you'd probably end up with something like this:

```json
{
  "scripts": {
    "prisma": "dotenv -e .env.development prisma"
  }
}
```

And then `yarn prisma migrate dev`, `yarn prisma migrate reset`, etc. But now the env is hard coded and is confusing because you can't do `NODE_ENV=test yarn prisma migrate reset` for example.

---

Hopefully this shows how the core problem currently is the ORM dictating how env vars are loaded and not giving a way for us to decide how we want to load env vars. (see my previous post above)

---

For more of an example, here's what I have in a Blitz app except the prisma commands shown here are what I want if we get this env issue resolved. `NODE_ENV=test` will cause prisma to load `.env` and `.env.test`

```json
{
  "scripts": {
    "test:server": "blitz build && blitz start --production -p 3099",
    "test:e2e": "NODE_ENV=test start-server-and-test test:server http://localhost:3099 'cypress run'",
  }
}
```

And in `cypress/plugins/index.ts` I have essentially this code:

```ts
let dbSetup = false
const pluginConfig: Cypress.PluginConfig = (on, _config) => {
  on("task", {
    factory,
    "db:reset": async () => {
      if (!dbSetup) {
        try {
          // Only need to do this once at startup
          logger.info("Resetting database...")
          await command("yarn prisma migrate reset", { timeout: 15000 })
          logger.info("Database reset.")
          dbSetup = true
        } catch (error) {
          logger.error(
            "Failed to set up database"
          )
        }
      }
      return true
    },
    "db:clear": async () => {
      // Delete all data without dropping tables, so migration isn't required
      await db.$executeRaw(`
        do
        $$
        declare
          l_stmt text;
        begin
          select 'truncate ' || string_agg(format('%I.%I', schemaname, tablename), ',')
          into l_stmt from pg_tables
          where schemaname in ('public');
          execute l_stmt;
        end;
        $$
      `)
      return true
    },
    "db:seed": async () => {
      await seed()
      return true
    },
  })
}
```

And in `cypress/support/index.ts` I have:

```ts
before(() => {
  // Initial test DB migration/reset
  cy.task("db:reset")
})

beforeEach(() => {
  // Clear and seed test db before each test
  cy.task("db:clear")
  cy.task("db:seed")
})
```

**Comment by narze** [comment]
I wrapped @nephix's solution to a script and reuse it with multiple commands, but doing this way you have to run `yarn prisma` or `npm run prisma` instead of `npx prisma`

```
// Wrapped prisma with dotenv-flow-cli, also ensure that NODE_ENV is presen
"prisma": "NODE_ENV=${NODE_ENV:-development} dotenv-flow -- prisma",

"migrate:dev": "yarn prisma migrate dev",
"migrate:dev:create": "yarn prisma migrate dev --create-only",
"migrate:deploy": "yarn prisma migrate deploy",
"migrate:reset": "yarn prisma migrate reset",
"migrate:resolve": "yarn prisma migrate resolve",
"prisma:generate": "yarn prisma generate",
"prisma:generate:watch": "yarn prisma generate --watch",
"prisma:studio": "yarn prisma studio",
```

**Comment by oncet** [comment]
Does this affect during app runtime or am I missing something? I'm having problem getting Prisma to read the `DATABASE_URL` from a `.env.local` file on a Next.js project.

**Comment by kleinfreund** [comment]
**TL;DR**: I’d like to run `npx prisma db pull --env .env.development` (or any prisma CLI command that relies on the environment) so that it loads the environment from .env.development instead of .env.

---

> […], the part I don't understand is that `prisma migrate dev` by itself cannot know to switch environments for migrating the test or production database. You need to indicate that somehow, maybe:
> 
>     * `prisma migrate dev --env .env.production`

The prisma CLI already makes a choice: It attempts to load the environment from an .env file if present. When I’m setting up prisma using a guide like [Start from scratch](https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch-typescript-postgres/), I’m immediately hitting a road block during the `npx prisma init` instructions. Specifically, when one runs `npx prisma db pull`, the script fails on account of `DATABASE_URL` not being present in the environment. In my existing application, the local or development environment happens to be defined in .end.development, not .env.

Now the only workarounds for this, as outlined in the initial post, are not appealing to me. I don’t want to introduce another way to specify environments in my application. I already have that built into it using `dotenv`. Having two ways of doing the same thing is more brittle than should be necessary. I also don’t want to introduce another tool for something prisma can already do: it is already able to load .env files. It just doesn’t expose a way to let me choose from _which_ file it loads from. Sure, I could install dotenv-cli or something of the likes to manually load the environment, but that’s one more layer of environment handling existing in my application in addition to the existing ones: the one I already had and the one internally used by prisma. Or I change my application to use the .env file instead of .env.development. I’m not particularly thrilled about that either because it’s a deliberate choice that I call this file .env.development. It’s specific and it doesn’t give off the impression that it’s just a base environment that’s meant to merged with another. It’s not. I’d like to keep it. Also, if another tools comes along that specifically can’t load from .env because it too makes a choice just like prisma does but one that doesn’t happen to align with prisma’s choice, all bets are off and I *must* add extra tooling around something *both* tools already do well on their own.

**Comment by akomm** [comment]
For those who don't want to add ~4 more dependencies to get dotenv-flow-cli. Create a file, for example `.scripts/prisma.js`:
```js
"use strict"

process.chdir(__dirname + "/..")

const args = process.argv.slice(2)
const nodeEnv = args.shift()
const envs = {dev: "development", prod: "production"}

if (!Object.keys(envs).includes(nodeEnv)) {
  process.stderr.write("Usage: prisma [dev|prod] [prisma-commands]\n")
  process.exit(0)
}

process.env["NODE_ENV"] = envs[nodeEnv]

require("dotenv-flow").config()

const {spawn} = require("child_process")
const opts = {stdio: "inherit"}

spawn("node_modules/.bin/prisma", args, opts)
```

And with this in `package.json` `scripts`:
```json
{"prisma": "node .scripts/prisma.js"}
```
Run on command line anything you want, for example: `yarn prisma dev db push` (or npm...)

If you forget to specify env, you get a notification.

I like the explicit way of doing it and if I forget, I will be notified. And its not that hard writing dev or prod and knowing what you want.

**Comment by flybayer** [comment]
Update on Next.js/Blitz env variable loading. There's an RFC in process that looks likely to be at least partially adopted: https://github.com/vercel/next.js/discussions/25764

The TLDR is that it will add a `-e` flag such that `next -e staging` would load `.env.staging`, `.env.staging.local`, `.env.local`, and `.env`. Also it would support an APP_ENV env var such that `APP_ENV=staging next` would be the same as `next -e staging`.

So perhaps we can add a config item to prisma schema to set which env variable will be used to set the environment. That way folks who want NODE_ENV can use that or folks who want APP_ENV can use that too.

**Comment by dthomason** [comment]
I just have to say, I really like prisma so far since the past few weeks of using it.  But this is my biggest gripe.  I feel like a lot of time could have been saved by just supporting what every other standard ORM supports.  All of these fixes seem  a bit hacky to me, and for integration tests having to use docker seems like overkill and way too much time wasted building up and tearing down.  I get the sense the Prisma team is not too big on TDD otherwise this would be a no brainer.

**Comment by matthewmueller** [comment]
> I feel like a lot of time could have been saved by just supporting what every other standard ORM supports.

Do you have an ORM in mind? I'd love to hear which ones support this out of the box.

> I get the sense the Prisma team is not too big on TDD otherwise this would be a no brainer.

I'm not sure I understand this. 

For integration tests, isn't it the responsibility of the test runner like Mocha or Jest to make sure you've loaded your test environment? 

Prisma seems like one small piece of the test harness, where you'd do something like: `await execa("prisma", ["migrate"], { env: process.env })`, where `process.env` is populated from `.env` via Mocha, Jest or manually with dot-env.

**Comment by Knivemare** [comment]
> Currently we only have two options:
> 
> * load env variables before invoking prisma
> * use `.env` or `.env.local`

How can '.env.local' be used?

**Comment by Radmehr-aghdam** [comment]
> > Currently we only have two options:
> > 
> > * load env variables before invoking prisma
> > * use `.env` or `.env.local`
> 
> How can '.env.local' be used?

https://github.com/prisma/prisma/issues/3865#issuecomment-767665090

**Comment by ZetiMente** [comment]
This would be a nice feature to have with VITE Modes as well

**Comment by Jolg42** [comment]
@ZetiMente I guess you mean this

> By default, the dev server (dev command) runs in development mode and the build command run in production mode.
This means when running vite build, it will load the env variables from .env.production if there is one
...
https://vitejs.dev/guide/env-and-mode.html#modes

**Comment by ZetiMente** [comment]
@Jolg42 Yes, I already have env/.env/.development.local, env/.env.development and env/.env.production files that work with those Vite modes. It would be nice to have a VITE_PRISMA_DATABASE_URL or just PRISMA_PRISMA_DATABASE_URL in these files instead of having a .env file just for Prisma.

**Comment by hyoretsu** [comment]
I just want to use Prisma Data Proxy and create my migrations

**Comment by janpio** [comment]
If you still need to load the values into your environment manually, then `dotenv-cli` or a similar package could be used and called before you execute either the npm script or you could put it into the npm script itself.
